### DISCUSSION TITLE
smaller jar with fewer dependencies?

### DISCUSSION THREAD

**Discussion by goodale** [nielsbasjes/yauaa#1168] [Q&A] [answered]
I noticed the jar file is over 2MB.   This seems very large for a parser.  So not only increases war file size but also presumably adds additional vectors for security issues.

Is it possible to build a slim version that can be used in place of the older `uadetector` or `uap-java`, both of which are much smaller?

**Comment by nielsbasjes** [comment]
I have tried to minimise the dependencies and I have stripped as many of the unneeded classes as possible. So this os as small as it gets.

After startup this system uses a few hundred MiB of RAM, also a lot more than the others. And all of this makes the real goal possible: Best possible result in a time as short as possible.

So my main question is why is a 2MiB library a problem for you?

**Reply by nielsbasjes** [reply, marked as answer]
I pulled the jar apart and did some stats on the compressed content for version 7.22.0.

Summary: There is no way to reduce the size any further.

Group | Compressed bytes | Percentage of compressed file | Purpose 
-- | --: | --: | --
META-INF/ | 4614 | 0.2% | Needed in any jar
nl/basjes/parse/useragent/ | 402560 | 19.5% | The analyzer engine 
nl/basjes/parse/useragent/parser/ | 224837 | 10.9% | The generated useragent parser
nl/basjes/parse/useragent/utils/publicsuffixlist/ | 7802 | 0.4% | Handling the domain suffixes
nl/basjes/parse/useragent/utils/springframework/ | 103203 | 5.0% | Dynamically finding the available yaml files
nl/basjes/shaded/org/antlr/ | 256797 | 12.4% | Runtime for the generated parser
nl/basjes/shaded/org/yaml/snakeyaml | 273181 | 13.2% | Loading the yaml files.
UserAgents/ | 795374 | 38.5% | The yaml files with all the pattern rules.
Total Result | 2068368 | 100.0%

**Reply by goodale** [reply]
thank you.  I see a large part of the bundle is the yaml file, which makes sense.   With some enterprise apps having upwards of hundreds of dependencies it is just prudent to use libs that are as small as possible, both to minimize start-up time but also publish time, etc.   This is 
 a library that should be fine to filter from TLD scanning so that is one area to improve scan time.   Thanks again.
